src/miniscript/musig_key.rs

```diff
@@ -0,0 +1,269 @@
+    /// Returns an XOnlyPublicKey from a KeyExpr
+    pub fn key_agg(&self) -> bitcoin::XOnlyPublicKey {
+        match self {
+            KeyExpr::<Pk>::SingleKey(pk) => pk.to_x_only_pubkey(),
+            KeyExpr::<Pk>::MuSig(_keys) => {
+                let secp = Secp256k1::verification_only();
+                self.key_agg_helper(&secp)
+            }
+        }
+    }
+
+    #[cfg(feature = "std")]
+    fn key_agg_helper(&self, secp: &Secp256k1<VerifyOnly>) -> bitcoin::XOnlyPublicKey {
+        match self {
+            KeyExpr::<Pk>::SingleKey(pk) => pk.to_x_only_pubkey(),
+            KeyExpr::<Pk>::MuSig(keys) => {
+                let xonly = keys
+                    .iter()
+                    .map(|key| key.key_agg_helper(secp))
+                    .collect::<Vec<_>>();
+                let key_agg_cache = MusigKeyAggCache::new(secp, &xonly);
+                key_agg_cache.agg_pk()
+            }
+        }
+    }
+
+    #[cfg(not(feature = "std"))]
+    fn key_agg_helper(&self, _secp: &Secp256k1<VerifyOnly>) -> bitcoin::XOnlyPublicKey {
+        match self {
+            KeyExpr::<Pk>::SingleKey(pk) => pk.to_x_only_pubkey(),
+            KeyExpr::<Pk>::MuSig(_keys) => {
+                unimplemented!("Musig not supported for no-std");
+            }
+        }
+    }
+}
+
+impl<Pk: MiniscriptKey> ForEachKey<Pk> for KeyExpr<Pk> {
+    fn for_each_key<'a, F: FnMut(&'a Pk) -> bool>(&'a self, mut pred: F) -> bool
+    where
+        Pk: 'a,
+        Pk::RawPkHash: 'a,
+    {
+        let keys_res = self.iter().all(|key| pred(key));
+        keys_res
+    }
+}
+
+impl<P, Q> TranslatePk<P, Q> for KeyExpr<P>
+where
+    P: MiniscriptKey,
+    Q: MiniscriptKey,
+{
+    type Output = KeyExpr<Q>;
+    fn translate_pk<T, E>(&self, t: &mut T) -> Result<Self::Output, E>
+    where
+        T: Translator<P, Q, E>,
+    {
+        match self {
+            KeyExpr::<P>::SingleKey(pk) => Ok(KeyExpr::SingleKey(t.pk(pk)?)),
+            KeyExpr::<P>::MuSig(vec) => {
+                let mut new_vec: Vec<KeyExpr<Q>> = vec![];
+                for x in vec {
+                    new_vec.push(x.translate_pk(t)?)
+                }
+                Ok(KeyExpr::MuSig(new_vec))
+            }
+        }
+    }
+}
+
+impl<Pk: MiniscriptKey> KeyExpr<Pk> {
+    /// Returns the Pk if KeyExpr is SingleKey, otherwise None
+    pub fn single_key(&self) -> Option<&Pk> {
+        match self {
+            KeyExpr::<Pk>::SingleKey(ref pk) => Some(pk),
+            KeyExpr::<Pk>::MuSig(_) => None,
+        }
+    }
+
+    /// Return vector of bytes, in case of musig, it is first aggregated
+    /// and then converted to bytes
+    pub fn to_vec(&self) -> Vec<u8>
+    where
+        Pk: ToPublicKey,
+    {
+        match self {
+            KeyExpr::<Pk>::SingleKey(ref pk) => pk.to_public_key().to_bytes(),
+            KeyExpr::<Pk>::MuSig(_) => {
+                let agg_key = self.key_agg();
+                agg_key.to_public_key().to_bytes()
```

Comment:
You should not need the .to_public_key here

Reply:
Ok, I have removed `to_public_key()` and used the `serialize()` function.